server/app/filters/UnsupportedBrowserFilter.java

```diff
@@ -0,0 +1,39 @@
+package filters;
+
+import java.util.Optional;
+import play.libs.streams.Accumulator;
+import play.mvc.Call;
+import play.mvc.EssentialAction;
+import play.mvc.EssentialFilter;
+import play.mvc.Http;
+import play.mvc.Results;
+
+/**
+ * Filter that redirects requests from unsupported browsers (e.g. IE11) to a page that tells user to
+ * use newer browser.
+ */
+public class UnsupportedBrowserFilter extends EssentialFilter {
+  @Override
+  public EssentialAction apply(EssentialAction next) {
+    return EssentialAction.of(
+        request -> {
+          if (UnsupportedBrowserFilter.isUnsupportedBrowser(request)) {
+            Call unsupportedBrowser =
+                controllers.routes.SupportController.handleUnsupportedBrowser();
+            // Don't redirect if request is for a static asset or for the unsupported browser page
+            // itself.
+            if (!request.path().startsWith("/assets")
+                && !request.path().startsWith(unsupportedBrowser.path())) {
+              return Accumulator.done(Results.redirect(unsupportedBrowser));
+            }
+          }
+          return next.apply(request);
+        });
+  }
+
+  private static boolean isUnsupportedBrowser(Http.RequestHeader request) {
+    Optional<String> userAgent = request.getHeaders().get(Http.HeaderNames.USER_AGENT);
+    // Trident/[5-7].0 browsers are IE9-IE11. Everything else consider supported.
```

Comment:
Blocking: User agent matching tends to lead to a lot of false positives / false negatives. If possible, I'd consider checking to see if there are any libraries out there that handle this and making use of them.

That said, the regex here is straightforward. I just worry about what needs to be matched changing over time and us forgetting to update, which would lead to an unsupported browser page even if we actually did support it :(.

Reply:
I looked at some libraries, but couldn't find something suitable for java. I found this: https://github.com/nielsbasjes/yauaa but i had few issues:
1. Required adding a bunch of logging libraries like log4j or slf4j. I don't fully understand logging landscape in java to evaluate what it might mean for the app.
2. Required non-trivial initialization. You need to initialize instance to parse UA string and it says it may take some time and they warn against using static members: [doc](https://yauaa.basjes.nl/using/#avoid-using-it-as-a-static-member). So would need to do something around singleton injection...

Generally wasn't sure how fast the UA parsing/detection would be when using full-blown library. And here we are more sensitive to latency as this code runs for every single request. Hm, talking about that, I should pre-compile regex.

I agree that if we start excluding more browsers and this logic starts to grow - we should use some library.